### ISSUE TITLE
Blend issue with internal lightmap

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#326]
Initially reported there: https://github.com/Unvanquished/Unvanquished/issues/782
But the thread reported two issues, this one if for the blending issue.

Here is an example of that issue (taken from Unvanquished/Unvanquished#782 thread):

internal:  
[![lightmap](http://dl.illwieckz.net/b/unvanquished/bugs/lightmaps/unvanquished_2015-07-30_005613_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/lightmaps/unvanquished_2015-07-30_005613_000.jpg)

external:  
[![lightmap](http://dl.illwieckz.net/b/unvanquished/bugs/lightmaps/unvanquished_2015-07-30_005923_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/lightmaps/unvanquished_2015-07-30_005923_000.jpg)

The depicted map is the Terminus map.

Released maps usually don't reproduce the bug since we usually use external lightmaps as they compress better. This issue may still bite us in the back later, and we can't assume any Dæmon-based game will repackage every legacy map available on this world, and maybe some may not have the right to.

**Comment by illwieckz** [comment]
While working on #301 (_merge lightmap and vertex (lightgrid) world and entity lighting_) and especially on the workaround for #324 (_light color restoration algorithm produces awful results when using lightdir from lightgrid_) I mistakenly reproduced this bug while the lightmaps were external.

The bug may be related to #302 (_need a reliable way to detect if a given surface is one of a bsp model_) since I suspect I reproduced the bug by making the non-lightmap code use the `color.rgb = vec3(0.0);` snippet usually used for lightmap only (see #320: _Fixing the lights, something does not look legit in the code_), maybe those surfaces are not lightmapped. But that's not enough since we still don't know why internal and external lightmap would not trigger the same code route.

**Comment by illwieckz** [comment]
When #301 is applied, It behaves the opposite:

internal:
[![lightmap dark blend](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192540_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192540_000.jpg)

external:
[![lightmap dark blend](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192726_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192726_000.jpg)

 Though it looks better in “worst case”.

This is the related shader:

```
textures/terminus/mlabpub01
{
	qer_editorimage textures/terminus/mlabpub01
	q3map_surfacelight 800
	q3map_lightsubdivide 128
	surfaceparm trans
	surfaceparm nonsolid
	surfaceparm nomarks
	surfaceparm lightfilter
	nopicmip
	cull none

	{
		map $lightmap
		rgbGen identity
	}
	{
		map textures/terminus/mlabpub01
		rgbGen identity
		blendFunc GL_DST_COLOR GL_ZERO
	}
	{
		map textures/terminus/neonlightpub
		blendfunc GL_ONE GL_ONE
	}
}
```

We note it's a texture applied on a lightmap (not the opposite) with a glow map.

**Comment by illwieckz** [comment]
We also get this, once #301 is merged:

internal:

[![lightmap dark blend](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192527_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192527_000.jpg)

external:

[![lightmap dark blend](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192358_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/lightmap-dark-blend/unvanquished_2020-04-23_192358_000.jpg)

This looks similar to #322 (_broken lantern/candle in metro map_), this is the shader:

```
textures/terminus/embers
{
	qer_editorimage textures/terminus/embers01
	nopicmip
	surfaceparm trans
	surfaceparm nomarks
	surfaceparm nolightmap
	{
		map textures/terminus/embers01
		blendfunc gl_dst_color gl_src_color
	}
	{
		map textures/terminus/embers02
		blendfunc add
		rgbgen wave sin .3 .1 0.3 1
		alphaGen wave sin .3 .1 0.3 1
	}
	{
		map $lightmap
		blendfunc filter
	}
}
```

**Comment by illwieckz** [comment]
What is interesting is that external lightmaps are loaded one by one this way:

```
] imagelist
[…]
 102:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0000.webp 
 103:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0001.webp 
 104:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0002.webp 
 105:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0003.webp 
 106:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0004.webp 
 107:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0005.webp 
 108:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0006.webp 
 109:  128  128  no    2D   RGB8     s.clmp  t.clmp   maps/terminus/lm_0007.webp 
[…]
```

While internal lightmaps are padded in a single texture this way:

```
] imagelist
[…]
 102: 2048 2048  yes   2D   RGB8     s.clmp  t.clmp   _fatlightmap0 
[…]
```

So there is probably many places to introduce different behaviour. I'm would recommend to load each internal lightmap as one texture in a way we always run the same code whatever the way it is stored.

**Comment by illwieckz** [comment]
I managed to reproduce the initial issue by using a black deluxe map, and current issue seems to occurs when there light grid is loaded instead of a white light map.

The issue seems to be that if lightmapping is disabled on a surface, lightmap code must still run (instead of lightgrid one) with a white image.

I also noticed current master did not have the first issue (probably thanks to #309: _tell a surface belongs to the bsp, properly select the renderer_). So that was likely an issue related to the fact that when the code for grid lighting and light map lighting is merged, we have to take care of the fact the light grid must not be used if there is no lightmap but light mapping is disabled. Because of the design, no lightmap use lightmap code but with a white image.

**Comment by illwieckz** [comment]
I don't have a clear explanation about the fact the behaviour was not the same either the lightmap is internal or not (I would expect the code to behave wrongly in all cases). So there may be dragons hiding in the corner, I have some tracks to follow it but that dragon will be blindly killed without approaching it the day of we get rid of the `fatLightmap` mechanism to factorize the lightmap code and fix #327 (_spilling out internal lightmap_).
